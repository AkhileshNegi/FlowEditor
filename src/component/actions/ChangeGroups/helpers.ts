import { Group, ChangeGroups } from '../../../flowTypes';
import { SearchResult } from '../../../store';

export const mapGroupsToSearchResults = (groups: Group[]) =>
    groups.map(({ name, uuid }) => ({ name, id: uuid }));

export const mapSearchResultsToGroups = (searchResults: SearchResult[]) =>
    searchResults.map(result => ({
        uuid: result.id,
        name: result.name
    }));

export const getGroups = (action: ChangeGroups): SearchResult[] => {
    if (action.groups == null) {
        return [];
    }

    if (action.groups.length && action.type !== 'add_contact_groups') {
        return mapGroupsToSearchResults(action.groups);
    }

    return [];
};
