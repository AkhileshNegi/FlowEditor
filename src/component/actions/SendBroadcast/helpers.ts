import { BroadcastMsg, Contact, Group } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';

export const getRecipients = (action: BroadcastMsg): Asset[] => {
    const selected = action.groups.map((group: Group) => {
        return { id: group.uuid, name: group.name, type: AssetType.Group };
    });

    return selected.concat(
        action.contacts.map((contact: Contact) => {
            return { id: contact.uuid, name: contact.name, type: AssetType.Contact };
        })
    );
};
