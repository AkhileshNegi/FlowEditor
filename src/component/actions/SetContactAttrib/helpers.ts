import { Types } from '../../../config/typeConfigs';
import { Action, SetContactField, SetContactName, SetContactProperty } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';
import { snakify, titleCase } from '../../../utils';

export type ContactProperty = Types.set_contact_name;

export const newFieldAction = ({
    uuid,
    value,
    name
}: {
    uuid: string;
    value: string;
    name: string;
}): SetContactField => ({
    type: Types.set_contact_field,
    field: {
        key: snakify(name),
        name: titleCase(name)
    },
    uuid,
    value
});

export const newPropertyAction = ({
    uuid,
    value,
    type
}: {
    uuid: string;
    value: string;
    type: AssetType;
}): SetContactProperty => {
    const action: Partial<Action> = {
        uuid
    };
    switch (type) {
        case AssetType.Name:
            return {
                ...action,
                type: Types.set_contact_name,
                name: value
            } as SetContactName;
    }
};

export const fieldToAsset = ({ field: { key, name } }: SetContactField): Asset => ({
    id: key,
    name,
    type: AssetType.Field
});

export const propertyToAsset = (property: SetContactProperty): Asset => {
    switch (property.type) {
        case Types.set_contact_name:
            return {
                type: AssetType.Name,
                name: titleCase(AssetType.Name),
                id: AssetType.Name
            };
    }
};
