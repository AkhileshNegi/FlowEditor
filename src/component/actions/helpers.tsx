import * as React from 'react';

import { Asset, AssetType } from '../../services/AssetService';

const styles = require('../shared.scss');

export const renderAssetList = (assets: Asset[], max: number): JSX.Element[] => {
    return assets.reduce((elements, asset, idx) => {
        if (idx <= max - 1) {
            elements.push(renderAsset(asset));
        } else if (idx > max - 1 && elements.length === max) {
            elements.push(<div key="ellipses">...</div>);
        }
        return elements;
    }, []);
};

export const renderAsset = (asset: Asset) => {
    switch (asset.type) {
        case AssetType.Group:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeGroup} icn-group`} />
                    {asset.name}
                </div>
            );
        case AssetType.Label:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeLabel} icn-label`} />
                    {asset.name}
                </div>
            );
    }

    return (
        <div className={styles.nodeAsset} key={asset.id}>
            {asset.name}
        </div>
    );
};
