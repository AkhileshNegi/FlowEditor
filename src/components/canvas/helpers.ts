import { FlowPosition } from '~/flowTypes';
import { CanvasPositions } from '~/store/editor';

export const collides = (a: FlowPosition, b: FlowPosition) => {
    // don't bother with collision if we don't have full dimensions
    /* istanbul ignore next */
    if (!a.bottom || !b.bottom) {
        return false;
    }

    return !(b.left > a.right || b.right < a.left || b.top > a.bottom || b.bottom < a.top);
};

export const getCollisions = (
    positions: CanvasPositions,
    box: FlowPosition
): { [uuid: string]: FlowPosition } => {
    const collisions = {};
    for (const nodeUUID of Object.keys(positions)) {
        const position = positions[nodeUUID];
        if (collides(box, position)) {
            collisions[nodeUUID] = position;
        }
    }
    return collisions;
};
