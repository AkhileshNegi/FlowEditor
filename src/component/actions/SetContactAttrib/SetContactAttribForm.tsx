import * as React from 'react';
import { ConfigProviderContext } from '../../../config';
import { Types } from '../../../config/typeConfigs';
import { SetContactAttribute, SetContactField, SetContactProperty } from '../../../flowTypes';
import ConnectedAttribElement from '../../form/AttribElement';
import ConnectedTextInputElement from '../../form/TextInputElement';
import { newFieldAction, newPropertyAction, fieldToAsset, propertyToAsset } from './helpers';
import { fakePropType } from '../../../config/ConfigProvider';
import AssetService, { Asset, AssetType } from '../../../services/AssetService';

export interface SetContactAttribFormProps {
    action: SetContactAttribute;
    onBindWidget: (ref: any) => void;
    updateAction: (action: SetContactAttribute) => void;
}

export const ATTRIB_HELP_TEXT =
    'Select an existing attribute to update or type any name to create a new one';
export const TEXT_INPUT_HELP_TEXT =
    'The value to store can be any text you like. You can also reference other values that have been collected up to this point by typing @run.results or @webhook.json.';

export default class SetContactAttribForm extends React.Component<SetContactAttribFormProps> {
    public static contextTypes = {
        endpoints: fakePropType,
        assetService: fakePropType
    };

    constructor(props: SetContactAttribFormProps, context: ConfigProviderContext) {
        super(props);
        this.onValid = this.onValid.bind(this);
    }

    public onValid(widgets: { [name: string]: any }): void {
        const { wrappedInstance: { state: { attribute } } } = widgets.Attribute;
        const { wrappedInstance: { state: { value } } } = widgets.Value;

        if (attribute.type === AssetType.Field) {
            // include our contact field in our local storage
            const assetService: AssetService = this.context.assetService;
            assetService.getFieldAssets().add(attribute);
            this.props.updateAction(newFieldAction(this.props.action.uuid, value, attribute.name));
        } else {
            this.props.updateAction(
                newPropertyAction(this.props.action.uuid, value, attribute.name)
            );
        }
    }

    private getInitial(): Asset {
        if (this.props.action.type === Types.set_contact_field) {
            return fieldToAsset(this.props.action as SetContactField);
        } else {
            return propertyToAsset(this.props.action as SetContactProperty);
        }
    }

    public render(): JSX.Element {
        const initial = this.getInitial();
        return (
            <>
                <ConnectedAttribElement
                    ref={this.props.onBindWidget}
                    name="Attribute"
                    showLabel={true}
                    assets={this.context.assetService.getFieldAssets()}
                    helpText={ATTRIB_HELP_TEXT}
                    initial={initial}
                    add={true}
                    required={true}
                />
                <ConnectedTextInputElement
                    ref={this.props.onBindWidget}
                    name="Value"
                    showLabel={true}
                    value={this.props.action.value}
                    helpText={TEXT_INPUT_HELP_TEXT}
                    autocomplete={true}
                />
            </>
        );
    }
}
