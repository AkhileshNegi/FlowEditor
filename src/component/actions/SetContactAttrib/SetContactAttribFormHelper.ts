import { FormHelper, Types } from '../../../config/typeConfigs';
import {
    Channel,
    SetContactAttribute,
    SetContactChannel,
    SetContactField,
    SetContactLanguage,
    SetContactName,
} from '../../../flowTypes';
import { removeAsset } from '../../../services/AssetService';
import {
    NodeEditorSettings,
    SetContactAttribFormState,
    SetContactChannelFormState,
    SetContactFieldFormState,
    SetContactLanguageFormState,
    SetContactNameFormState,
} from '../../../store/nodeEditor';
import { getLanguage } from '../../../utils/languageMap';
import { assetToField, channelToAsset, fieldToAsset, languageToAsset, propertyToAsset } from './helpers';

export type SetContactAttribFormHelperActionTypes =
    | Types.set_contact_field
    | Types.set_contact_name
    | Types.set_contact_language
    | Types.set_contact_channel;

export class SetContactAttribFormHelper implements FormHelper {
    // passed an existing action or null
    public initializeForm(
        settings: NodeEditorSettings,
        actionType: SetContactAttribFormHelperActionTypes
    ): SetContactAttribFormState {
        let formState: SetContactAttribFormState;

        // if we have an existing contact attribute action, use it
        if (settings.originalAction) {
            const action = settings.originalAction as SetContactAttribute;
            switch (action.type) {
                case Types.set_contact_field:
                    formState = {
                        field: { value: fieldToAsset(action.field) },
                        value: { value: action.value },
                        valid: true
                    } as SetContactFieldFormState;
                    break;
                case Types.set_contact_name:
                    formState = {
                        name: { value: propertyToAsset(Types.set_contact_name) },
                        value: { value: (action as SetContactName).name },
                        valid: true
                    } as SetContactNameFormState;
                    break;
                case Types.set_contact_language:
                    const { language } = action as SetContactLanguage;
                    formState = {
                        language: { value: propertyToAsset(Types.set_contact_language) },
                        value: {
                            value: language
                                ? languageToAsset(
                                      getLanguage((action as SetContactLanguage).language)
                                  )
                                : removeAsset
                        },
                        valid: true
                    } as SetContactLanguageFormState;
                    break;
                case Types.set_contact_channel:
                    const { channel } = action as SetContactChannel;
                    formState = {
                        channel: { value: propertyToAsset(Types.set_contact_channel) },
                        value: {
                            value: Object.keys(channel).length
                                ? channelToAsset((action as SetContactChannel).channel)
                                : removeAsset
                        },
                        valid: true
                    } as SetContactChannelFormState;
                    break;
            }
        } else {
            // otherwise, provide initial form state from scratch
            switch (actionType) {
                case Types.set_contact_field:
                    formState = {
                        field: { value: fieldToAsset() },
                        value: { value: '' },
                        valid: false
                    } as SetContactFieldFormState;
                    break;
                case Types.set_contact_name:
                    formState = {
                        name: { value: propertyToAsset(Types.set_contact_name) },
                        value: { value: '' },
                        valid: false
                    } as SetContactNameFormState;
                    break;
                case Types.set_contact_language:
                    formState = {
                        language: { value: propertyToAsset(Types.set_contact_language) },
                        value: { value: removeAsset },
                        valid: false
                    } as SetContactLanguageFormState;
                    break;
                case Types.set_contact_channel:
                    formState = {
                        channel: { value: propertyToAsset(Types.set_contact_channel) },
                        value: { value: removeAsset },
                        valid: false
                    } as SetContactChannelFormState;
                    break;
            }
        }

        return formState;
    }

    public stateToAction(
        actionUUID: string,
        formState: SetContactAttribFormState,
        formType: Types
    ): SetContactAttribute {
        let action: Partial<SetContactAttribute> = {
            uuid: actionUUID,
            type: formType
        };

        switch (formType) {
            case Types.set_contact_field:
                action = {
                    ...action,
                    field: assetToField((formState as SetContactFieldFormState).field.value),
                    value: (formState as SetContactFieldFormState).value.value
                } as SetContactField;
                break;
            case Types.set_contact_name:
                action = {
                    ...action,
                    name: formState.value.value
                } as SetContactName;
                break;
            case Types.set_contact_language:
                action = {
                    ...action,
                    // we return an empty string to indicate the value is being cleared
                    language:
                        (formState as SetContactLanguageFormState).value.value.id === removeAsset.id
                            ? ''
                            : (formState as SetContactLanguageFormState).value.value.id
                } as SetContactLanguage;
                break;
            case Types.set_contact_channel:
                action = {
                    ...action,
                    // we return an empty string to indicate the value is being cleared
                    channel:
                        (formState as SetContactChannelFormState).value.value.id === removeAsset.id
                            ? {}
                            : ({
                                  uuid: (formState as SetContactChannelFormState).value.value.id,
                                  name: (formState as SetContactChannelFormState).value.value.name
                              } as Channel)
                } as SetContactChannel;
                break;
        }

        return action as SetContactAttribute;
    }
}
