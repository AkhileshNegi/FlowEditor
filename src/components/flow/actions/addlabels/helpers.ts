import { AddLabelsFormState } from '~/components/flow/actions/addlabels/AddLabelsForm';
import { getActionUUID } from '~/components/flow/actions/helpers';
import { Types } from '~/config/interfaces';
import { AddLabels } from '~/flowTypes';
import { Asset, AssetType } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): AddLabelsFormState => {
    if (settings.originalAction && settings.originalAction.type === Types.add_input_labels) {
        const action = settings.originalAction as AddLabels;
        return {
            labels: {
                value: action.labels.map(label => {
                    return { id: label.uuid, name: label.name, type: AssetType.Label };
                })
            },
            valid: true
        };
    }

    return {
        labels: { value: [] },
        valid: false
    };
};

export const stateToAction = (
    settings: NodeEditorSettings,
    formState: AddLabelsFormState
): AddLabels => {
    return {
        type: Types.add_input_labels,
        labels: this.getAsset(formState.labels.value, AssetType.Label),
        uuid: getActionUUID(settings, Types.add_input_labels)
    };
};

export const getAsset = (assets: Asset[], type: AssetType): any[] => {
    return assets.filter((asset: Asset) => asset.type === type).map((asset: Asset) => {
        return { uuid: asset.id, name: asset.name };
    });
};
