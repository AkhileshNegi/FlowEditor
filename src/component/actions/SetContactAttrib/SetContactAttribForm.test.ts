import { getTypeConfig } from '../../../config';
import { Types } from '../../../config/typeConfigs';
import { composeComponentTestUtils, setMock } from '../../../testUtils';
import {
    createSetContactFieldAction,
    createSetContactNameAction
} from '../../../testUtils/assetCreators';
import ConnectedTextInputElement from '../../form/TextInputElement';
import { propertyToAsset } from './helpers';
import { SetContactAttribForm, SetContactAttribFormProps } from './SetContactAttribForm';
import { SetContactAttribFormHelper } from './SetContactAttribFormHelper';

const setContactNameAction = createSetContactNameAction();
const setContactFieldAction = createSetContactFieldAction();

const formHelper = new SetContactAttribFormHelper();

// starting w/ Types.set_contact_field props
const baseProps: SetContactAttribFormProps = {
    action: setContactFieldAction,
    formHelper,
    typeConfig: getTypeConfig(Types.set_contact_field),
    form: formHelper.actionToState(setContactFieldAction, Types.set_contact_field),
    onBindWidget: jest.fn(),
    updateAction: jest.fn(),
    updateSetContactAttribForm: jest.fn()
};

const { setup } = composeComponentTestUtils(SetContactAttribForm, baseProps);

describe(SetContactAttribForm.name, () => {
    describe('render', () => {
        it('should render self, children with base props', () => {
            const { wrapper, props } = setup();

            expect(wrapper).toMatchSnapshot();
            expect(wrapper.find(ConnectedTextInputElement).prop('value')).toBe(props.form.value);

            const setContactNameForm = formHelper.actionToState(
                setContactNameAction,
                Types.set_contact_name
            );

            // User changes attribute to Name
            wrapper.setProps({
                action: setContactNameAction,
                form: setContactNameForm,
                typeConfig: getTypeConfig(Types.set_contact_name)
            });

            expect(wrapper.find(ConnectedTextInputElement).prop('value')).toBe(
                setContactNameForm.value
            );
        });
    });

    describe('instance methods', () => {
        describe('onValid', () => {
            it('should call updateAction prop with new set_contact_field action', () => {
                const { wrapper, instance, props } = setup(true, {
                    updateAction: setMock()
                });

                instance.onValid();

                expect(props.updateAction).toHaveBeenCalledTimes(1);
                expect(props.updateAction).toHaveBeenCalledWith(
                    props.formHelper.stateToAction(
                        props.action.uuid,
                        props.form,
                        props.typeConfig.type
                    )
                );
            });
        });

        describe('handleAttribChange', () => {
            it('should call form-state-updater-thunk', () => {
                const { wrapper, instance, props } = setup(true, {
                    updateSetContactAttribForm: setMock()
                });
                const attribute = propertyToAsset(setContactNameAction);

                instance.handleAttribChange(attribute);

                expect(props.updateSetContactAttribForm).toHaveBeenCalledTimes(1);
            });
        });

        describe('handleValueChange', () => {
            it('should call form-state-updater-thunk', () => {
                const { wrapper, instance, props } = setup(true, {
                    updateSetContactAttribForm: setMock()
                });
                const mockEvent = {
                    currentTarget: {
                        value: 26
                    }
                };

                instance.handleValueChange(mockEvent);

                expect(props.updateSetContactAttribForm).toHaveBeenCalledTimes(1);
                expect(props.updateSetContactAttribForm).toHaveBeenCalledWith(
                    null,
                    mockEvent.currentTarget.value
                );
            });
        });
    });
});
