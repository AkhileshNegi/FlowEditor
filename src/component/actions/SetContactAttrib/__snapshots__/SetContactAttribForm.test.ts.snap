// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetContactAttribForm render should render self, children with base props 1`] = `
<React.Fragment>
  <Connect(AttribElement)
    add={true}
    assets={
      FieldAssets {
        "assetType": "field",
        "assets": Object {
          "name": Object {
            "id": "name",
            "name": "Name",
            "type": "name",
          },
        },
        "endpoint": "/assets/fields.json",
        "idProperty": "key",
        "localStorage": true,
      }
    }
    helpText="Select an existing attribute to update or type any name to create a new one"
    name="Attribute"
    onChange={[Function]}
    required={true}
    showLabel={true}
  />
  <Connect(TextInputElement)
    autocomplete={true}
    helpText="The value to store can be any text you like. You can also reference other values that have been collected up to this point by typing @run.results or @webhook.json."
    name="Value"
    onChange={[Function]}
    showLabel={true}
    value="25"
  />
</React.Fragment>
`;
